### PR TITLE
Check for pending IO in the portable thread pool's worker threads

### LINKED ISSUE #82207: Net 7 Kestrel windows service hangs after a period of time
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We just update our projects to .NET 7 from .NET 5 and are experiencing hangs where Kestrel quits responding.  It works for a day or so and then the windows service needs to be restarted.  This was not an issue on .NET5.  I gathered a memory dump to see if I can figure out what is causing it to hang.  Here are some screenshots.  This was gathered using ProcDump.

This API is installed at many unique locations and several have reported the same issue that they need to restart the service daily.  For some locations, we use Azure relay.  So far, the locations that go through the relay have not had any issues.

image

image

### Expected Behavior

The API should respond while running

### Steps To Reproduce

Let our project run for a day

### Exceptions (if any)

_No response_

### .NET Version

7.0.100

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by kouvel**
- When Resource Monitor is attached, some async IO operations are bound to the thread that issued it even though the IO handle is bound to an IOCP. If the thread exits, the async IO operation is aborted. This can lead to hangs or unexpected exceptions.
- Added a check that was missing in the portable thread pool implementation to prevent exiting a worker thread when it has pending IO

Fixes https://github.com/dotnet/runtime/issues/82207

### CODE REVIEW

**Comment by stephentoub** [comment]
> When Resource Monitor is attached, some async IO operations are bound to the thread that issued it even though the IO handle is bound to an IOCP. If the thread exits, the async IO operation is aborted. This can lead to hangs or unexpected exceptions.

Just to build confidence, do we know why this happens / can we find out why this happens?  Is it documented somewhere?

File: src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs [hunk 1]
```diff
@@ -115,6 +116,17 @@ private static void WorkerThreadStart()
                         }
                     }
 
+                    // The thread cannot exit if it has IO pending, otherwise the IO may be canceled
+                    bool success =
+                        Interop.Kernel32.GetThreadIOPendingFlag(
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:121]: Isn't this code meant to be portable across OSes?

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:121]: Ah yes, will fix

**Comment by kouvel** [comment]
> Just to build confidence, do we know why this happens / can we find out why this happens? Is it documented somewhere?

I don't know yet, still looking into it.

**Review by stephentoub — APPROVED** [review]

**Comment by kouvel** [comment]
Looks like the failures are unrelated.

**Comment by kouvel** [comment]
> Just to build confidence, do we know why this happens / can we find out why this happens? Is it documented somewhere?

There is a brief note in [AcceptEx docs](https://learn.microsoft.com/en-us/windows/win32/api/mswsock/nf-mswsock-acceptex):

> Note: All I/O initiated by a given thread is canceled when that thread exits. For overlapped sockets, pending asynchronous operations can fail if the thread is closed before the operations complete. See [ExitThread](https://learn.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-exitthread) for more information.

I'm not sure about the details though, I have a question out to Windows folks.

I'll go ahead and merge this for now, as it seems like something that should be fixed anyway.

**Comment by davidfowl** [comment]
We're patching this in 7.0 right?

**Comment by kouvel** [comment]
> We're patching this in 7.0 right?

I have PRs open for [7.0](https://github.com/dotnet/runtime/pull/82246) and [6.0](https://github.com/dotnet/runtime/pull/82248), they're currently under consideration for servicing.

**Comment by davidfowl** [comment]
cc @mconnew

**Comment by mconnew** [comment]
This looks like it has a pathological case where multiple IO operations can be pending for long periods of time with no real work to complete. This will cause a tight loop spinning CPU.  

The known scenario I'm concerned about is CoreWCF (or WCF client) which both tend to make IO calls on the IO thread pool. This is because after completing a pending receive, they send back an ACK and make the next pending receive call on the same thread that the completion happened on. The NetTcp transport has a default receive timeout of 10 minutes. For each active connection (which can be thousands in the server case, and in a 3-tier app the middle tier can have thousands of WCF clients) we have a pending receive.  

We had a customer issue with WCF where their employees would all log in to the client app within a short period of time (around 200,000 new client connections within about 15 minutes) which would have a burst of threads handling the connection handshakes. That case ended up being an lsass bottleneck, but their application would hit a problem with this implementation. In that scenario, you would have 200,000 pending IO connections across 400+ IO threads, so every IO thread has many pending IO operations. Once things settled down, the request rate was very low and the IO thread pool would shrink over time.  

The IO thread pool was able to shrink as new work was only getting executed via a call to GetQueuedCompletionStatus. The special OS behavior of threads bound to an IO completion port means a thread won't be preemptively context switched away from if it's not in an alertable state (doing actual work) which means all the extra IO threads calling GetQueuedCompletionStatus wouldn't end up running any new work. With this new implementation, that constraint doesn't exist as the work is dispatched to a non IOCP thread, which means the "extra" threads are still going to pick up "new" work, and that new work will be able to initiate new IO, and keeping the thread around. You could end up with hundreds of threads spinning consuming CPU, with any IO completions being evenly distributed between them. With enough concurrent connections, the threads will be self maintaining and never clean up, spending most of their time consuming CPU.

**Pull request merged**